new easyblock for condalock

JinyuHan99 opened the pull request:
(created using `eb --new-pr`)

JinyuHan99 commented:
Hi ocaisa,

Thank you for your review! I'm trying to modify this easyblock according to your wonderful suggestions.

Regarding "Use conda-lock to generate the lock file (may require platform specific support)", I couldn't find the part where EasyBuild could automatically recognise the system architechture. Could you please indicate where I can find this information or if I need to write my own logic to achieve that?

If I'm going to do it myself, I suppose I need to use an enum and transform them into the way conda-lock could recognize, right? (Or the laziest way is to let users specify the platform)

FYI, the conda-lock supported platforms look like this:
```
platforms:
  - linux-64
  - osx-64
  - win-64
  - osx-arm64  # For Apple Silicon, e.g. M1/M2
  - linux-aarch64  # aka arm64, use for Docker on Apple Silicon
  - linux-ppc64le
```

Regarding "Use the conda_cmd the user asked for to install the package with the conda-lock created lock file", if I understand correctly, I need to use `conda-lock install` to create the environment through a lock file, rather than using the `conda_cmd`. 

Thank you for your time and help!

All the best,
Leah

ocaisa commented on easybuild/easyblocks/generic/conda.py:
```diff
@@ -38,7 +39,7 @@
 
 
-class Conda(Binary):
+class CondaEnhance(Binary):
```
```suggestion
class Conda(Binary):
```
You need to retain the original class name

JinyuHan99 replied:
Hi ocaisa, I just fixed this class name error in the newest commit. Sorry about that.

ocaisa commented:
> Regarding "Use conda-lock to generate the lock file (may require platform specific support)", I couldn't find the part where EasyBuild could automatically recognise the system architechture. Could you please indicate where I can find this information or if I need to write my own logic to achieve that?

`get_system_info()` will give you more than you need, see https://github.com/easybuilders/easybuild-framework/blob/develop/easybuild/tools/systemtools.py#L1128

> Regarding "Use the conda_cmd the user asked for to install the package with the conda-lock created lock file", if I understand correctly, I need to use `conda-lock install` to create the environment through a lock file, rather than using the `conda_cmd`.
> 

I'm no expert here, but reading the documentation for `conda-lock`:
```bash
# Install conda-lock (should use conda/mamba as requested by the easyconfig, make sure version >= 1.0)
<conda/mamba> install --channel=conda-forge --name=<expected name> conda-lock
# identify the platform using `get_system_info` and create the necessary mapping (there are some useful options here)
conda-lock render -p <platform> (-c <channel> --no-dev-dependencies)
# NOTE: There is a chance that the above command will fail, so need to verify that it worked
# Now that we have a lock file we can use the requested tool for the installation
<conda/mamba> create -n <expected name> --file conda-<platform>.lock
```

JinyuHan99 commented:
Hi ocaisa,

Thanks for your review! This time I added a few things:

1. I realized there are two kinds of lock files: multi-platform and single-platform files. They need to be handled differently. 
2. Used `easybuild.tools.systemtools.get_system_info()` to get platform info.
3. Added checking if the platform appeared in lock file matches with the actual platform
4. Shipped the related files in installation.
5. Used conda cmd to create the environment at the end.

Please let me know if there's anything wrong or need to be improved. Really appreciate it.

All the best,
Leah

JinyuHan99 commented:
Hello,

The tests failed due to module yaml not found. Could you please indicate the right way to use yaml module in the test environment? Thanks!

All the best,
Leah

ocaisa commented:
@JinyuHan99 I'm on leave until September, won't be able to get back to this until then
